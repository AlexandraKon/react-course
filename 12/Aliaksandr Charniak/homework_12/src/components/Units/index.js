import { connect } from 'react-redux'

import Units from './Units'

const mapStateToProps = (state) => ({
  data: state.units,
})

export default connect(mapStateToProps)(Units)
