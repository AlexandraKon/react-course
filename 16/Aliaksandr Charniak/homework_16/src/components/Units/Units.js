import React, { useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Link } from 'react-router-dom'

import { getCorrectedArray } from '../../utils/utils'
import { ACTION_GET_UNITS_REQUESTED } from '../../redux/units/actions'

export const Units = (props) => {
  const { data } = props.data
  const { state } = props.location
  //const arr = data ? (data.units ? data.units : Array.isArray(data) ? data : [data]) : []
  const arr = getCorrectedArray(data, 'units')
  const dispatch = useDispatch()
  useEffect(() => {
    dispatch(ACTION_GET_UNITS_REQUESTED(state?.unitUrl))
  }, [dispatch, state?.unitUrl])
  return (
    <div>
      {arr &&
        arr?.map((item) => (
          <ul key={item.id}>
            <li>Id: {item.id}</li>
            <li>Name: {item.name}</li>
            <li>Description: {item.description}</li>
            <li>Expansion: {item.expansion}</li>
            <li>Age: {item.age}</li>
            <li>
              Created in:
              <Link
                to={{
                  pathname: `/units/${item.id}/created_in/${item.created_in.substring(
                    item.created_in.lastIndexOf('/') + 1
                  )}`,
                  state: {
                    structUrl: item.created_in,
                  },
                }}
              >
                {item.created_in.substring(item.created_in.lastIndexOf('/') + 1)}
              </Link>
            </li>
            <li>
              Cost:
              {Object.keys(item.cost).map((key, i) => {
                if (typeof item.cost[key] === 'object') {
                  return (
                    <ul key={i}>
                      {Object.keys(item.cost[key]).map((k, i) => (
                        <li key={i}>
                          {k}: {item.cost[key][k]}
                        </li>
                      ))}
                    </ul>
                  )
                } else {
                  return (
                    <ul key={i}>
                      <li>
                        {key}: {item.cost[key]}
                      </li>
                    </ul>
                  )
                }
              })}
            </li>
            <li>Build time: {item.build_time}</li>
            <li>Reload time: {item.reload_time}</li>
            <li>Attack Delay: {item.attack_delay}</li>
            <li>Movement Rate: {item.movement_rate}</li>
            <li>Line Of Sight: {item.line_of_sight}</li>
            <li>Hit points: {item.hit_points}</li>
            <li>Range: {item.range}</li>
            <li>Attack: {item.attack}</li>
            <li>Armor: {item.armor}</li>
            <li>Accuracy: {item.accuracy}</li>
            {item.attack_bonus && (
              <li>
                Attack Bonus:
                {item.attack_bonus.map((it, i) => (
                  <ul key={i}>
                    <li>{it}</li>
                  </ul>
                ))}
              </li>
            )}
            <li>Search Radius: {item.search_radius}</li>
            <li>Blast Radius: {item.blast_radius}</li>
            <li>Armor Bonus: {item.armor_bonus}</li>
          </ul>
        ))}
    </div>
  )
}

export default Units
