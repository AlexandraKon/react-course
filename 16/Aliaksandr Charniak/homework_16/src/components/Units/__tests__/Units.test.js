import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import configureMockStore from 'redux-mock-store'

import { Provider } from 'react-redux'

import Units from '..'

configure({ adapter: new Adapter() })

const mockStore = configureMockStore()
const store = mockStore({})

describe('<Units /> mounting behaviour', () => {
  const props = {
    data: {
      units: [
        {
          id: 1,
          name: 'Archer',
          description: 'Quick and light. Weak at close range; excels at battle from distance',
          expansion: 'Age of Kings',
          age: 'Feudal',
          created_in: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range',
          cost: {
            Wood: 25,
            Gold: 45,
          },
          build_time: 35,
          reload_time: 2.0,
          attack_delay: 0.35,
          movement_rate: 0.96,
          line_of_sight: 6,
          hit_points: 4,
          range: 30,
          attack: 4,
          armor: '0/0',
          accuracy: '80%',
        },
        {
          id: 2,
          name: 'Crossbowman',
          description: 'Upgraded archer',
          expansion: 'Age of Kings',
          age: 'Castle',
          created_in: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range',
          cost: {
            Wood: 25,
            Gold: 45,
          },
          build_time: 27,
          reload_time: 2.0,
          attack_delay: 0.35,
          movement_rate: 0.96,
          line_of_sight: 7,
          hit_points: 35,
          range: 5,
          attack: 5,
          armor: '0/0',
          attack_bonus: ['+3 spearmen'],
          accuracy: '85%',
        },
      ],
      technologies: 'https://age-of-empires-2-api.herokuapp.com/api/v1/technologies',
      structures: 'https://age-of-empires-2-api.herokuapp.com/api/v1/structures',
      civilizations: 'https://age-of-empires-2-api.herokuapp.com/api/v1/civilizations',
    },
    location: false,
  }

  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <Provider store={store}>
        <Units {...props} />
      </Provider>
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 1 <Units /> component', () => {
    expect(wrapper).toHaveLength(1)
  })
})
