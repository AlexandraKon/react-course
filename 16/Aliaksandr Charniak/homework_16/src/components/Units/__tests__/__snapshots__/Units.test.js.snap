// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Units /> mounting behaviour should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(Units)
    data={
      Object {
        "civilizations": "https://age-of-empires-2-api.herokuapp.com/api/v1/civilizations",
        "structures": "https://age-of-empires-2-api.herokuapp.com/api/v1/structures",
        "technologies": "https://age-of-empires-2-api.herokuapp.com/api/v1/technologies",
        "units": Array [
          Object {
            "accuracy": "80%",
            "age": "Feudal",
            "armor": "0/0",
            "attack": 4,
            "attack_delay": 0.35,
            "build_time": 35,
            "cost": Object {
              "Gold": 45,
              "Wood": 25,
            },
            "created_in": "https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range",
            "description": "Quick and light. Weak at close range; excels at battle from distance",
            "expansion": "Age of Kings",
            "hit_points": 4,
            "id": 1,
            "line_of_sight": 6,
            "movement_rate": 0.96,
            "name": "Archer",
            "range": 30,
            "reload_time": 2,
          },
          Object {
            "accuracy": "85%",
            "age": "Castle",
            "armor": "0/0",
            "attack": 5,
            "attack_bonus": Array [
              "+3 spearmen",
            ],
            "attack_delay": 0.35,
            "build_time": 27,
            "cost": Object {
              "Gold": 45,
              "Wood": 25,
            },
            "created_in": "https://age-of-empires-2-api.herokuapp.com/api/v1/structure/archery_range",
            "description": "Upgraded archer",
            "expansion": "Age of Kings",
            "hit_points": 35,
            "id": 2,
            "line_of_sight": 7,
            "movement_rate": 0.96,
            "name": "Crossbowman",
            "range": 5,
            "reload_time": 2,
          },
        ],
      }
    }
    location={false}
  />
</ContextProvider>
`;
